### DISCUSSION TITLE
How can I prevent mouse and touch inputs from interacting with world/entities underneath UI?

### DISCUSSION THREAD

**Discussion by miketwenty1** [bevyengine/bevy#14043] [Q&A]
I'm using NodeBundles/ButtonBundles etc..

If a user clicks on a button I'm also inadvertently capturing that click event for what is underneath the UI. I have a dynamic UI and a dynamic map underneath and would like to not have a hacky region where I ignore clicks. 

I used to do:
```
mut mouse: ResMut<ButtonInput<MouseButton>>,
mut touches: ResMut<Touches>,

....

mouse.clear();
touches.clear();
``` 

And then `.chain()` systems where my UI systems would precede my other systems, but it seems for Bevy 13.2, this doesn't seem to do the trick for me.

Is there an idiomatic way of accomplishing where if the UI is interacted with it clears any further mouse event?

**Comment by bugsweeper** [comment]
Please make some simple reproduction code, I placed two buttons at same place by position Absolute and this code:
```
// prints events as they come in
fn event_listener(mut events: EventReader<MouseButtonInput>) {
    for event in events.read() {
        info!("{event:#?}");
    }
}

// prints events resource if not empty
fn event_listener2(mut events: ResMut<ButtonInput<MouseButton>>) {
    if events.get_just_pressed().count() > 0 {
        info!("{events:#?}");
    }
}
```
gave output:
```
2024-06-27T10:08:15.322239Z  INFO test_event: MouseButtonInput {
    button: Left,
    state: Pressed,
    window: Entity {
        index: 0,
        generation: 1,
    },
}
2024-06-27T10:08:15.322239Z  INFO test_event: ResMut(
    ButtonInput {
        pressed: {
            Left,
        },
        just_pressed: {
            Left,
        },
        just_released: {},
    },
)
2024-06-27T10:08:15.438080Z  INFO test_event: MouseButtonInput {
    button: Left,
    state: Released,
    window: Entity {
        index: 0,
        generation: 1,
    },
}
```
As you see `ResMut<ButtonInput<MouseButton>>` gives just naked event without any entities, don't you iter entities sepparately of events?

**Comment by mockersf** [comment]
> Is there an idiomatic way of accomplishing where if the UI is interacted with it clears any further mouse event?

There's no notion of clearing events from one system for another.

What I did a few time is add a full screen button that is under the rest of UI, If that button get the click, then it's for the game. The [`FocusPolicy`](https://docs.rs/bevy/latest/bevy/ui/enum.FocusPolicy.html) component defines if a UI element should block or let click pass through

**Reply by miketwenty1** [reply]
@bugsweeper @mockersf 

This isn't for 2 UI buttons this would be for UI overlaying things underneath, like clicking a sprite. For 2 UI Buttons FocusPolicy::Block would be sufficient. 

Edit: Am I misunderstanding the FocusPolicy? Would it block a click for things other than other UI elements?

My hacky work around is to temporarily disable interaction with given entities if I set a UiInteraction Res to true and then after "release" of the input I turn UiInteraction to false and Sprites with a given component can be resumed to be interacted with.

**Reply by miketwenty1** [reply]
If there was some WorldFocusPolicy idea I think that would be the "easy street" to addressing my issue?

**Comment by claudijo** [comment]
@mockersf Could you explain in more detail on how you use a full screen button to conditionally prevent handling clicks in the world? 

Are you doing some variant of what @miketwenty1 suggests, that is

> My hacky work around is to temporarily disable interaction with given entities if I set a UiInteraction Res to true and then after "release" of the input I turn UiInteraction to false and Sprites with a given component can be resumed to be interacted with.

Or do you perform some other clever tricks using the full screen button?

**Reply by claudijo** [reply]
On a second thought, I guess you basically query for interactions on the full screen button to detect clicks in the window, instead of just checking for mouse button events in general(?)
